### PR TITLE
Multi point nms dev

### PR DESCRIPTION
**Pull request by Yipeng-Sun**
1.  Extend NMS operation from bounding box [x,y,h,w] to 4-points {x_i} i=1,...,4 or 16 points for curved text detection.
2. Note that the input params. support BBox, Quadrangle (4-point input), and 16-points input by judging the input format automatically.

### CODE REVIEW

File: paddle/fluid/operators/detection/multiclass_nms_op.cc [hunk 1]
```diff
@@ -42,18 +41,20 @@ class MultiClassNMSOp : public framework::OperatorWithKernel {
...
-    PADDLE_ENFORCE_EQ(box_dims[2], 4,
-                      "The 2nd dimension of Input(BBoxes) must be 4, "
-                      "represents the layout of coordinate "
-                      "[xmin, ymin, xmax, ymax]");
+    PADDLE_ENFORCE(box_dims[2] == 4 || box_dims[2] == 8 || box_dims[2] == 32,
+                   "The 2nd dimension of Input(BBoxes) must be 4 or 8, "
+                   "represents the layout of coordinate "
+                   "[xmin, ymin, xmax, ymax] or "
+                   "4 points: [x1, y1, x2, y2, x3, y3, x4, y4] or "
+                   "16 points: [xi, yi] i= 1,2,...,16");
```
Comment (zchengquan) [inline paddle/fluid/operators/detection/multiclass_nms_op.cc:51]: 这里是限制只能用4、8、16点吗？其他的偶数不适用吗？

Comment (Yipeng-Sun) [reply paddle/fluid/operators/detection/multiclass_nms_op.cc:51]: 支持固定偶数点设计，对常用8点、16点检测作为默认输入，不常用参数暂不需要开放。

File: paddle/fluid/operators/detection/gpc.cc [hunk 2]
```diff
@@ -0,0 +1,2191 @@
...
+  gpc_free<int>(p->hole);
+  gpc_free<gpc_vertex_list>(p->contour);
+  p->num_contours = 0;
+}
+
+void gpc_read_polygon(FILE *fp, int read_hole_flags, gpc_polygon *p) {
+  int c = 0;
+  int v = 0;
+
+  fscanf(fp, "%d", &(p->num_contours));
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/detection/gpc.cc:795]: 这里需要处理下`fscanf`的返回值，具体可以参考这里的讨论：https://stackoverflow.com/questions/34191613/warning-ignoring-return-value-of-fscanf-in-c

File: paddle/fluid/operators/detection/poly_util.cc [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+  Point_<T> point;
+  for (size_t i = 0; i < pts_num; i++) {
+    point.x = contour.vertex[i].x;
+    point.y = contour.vertex[i].y;
+    vec.push_back(point);
+  }
+}
+
+template <class T>
+T GetContourArea(std::vector<Point_<T>>& vec, bool oriented = false) {
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/detection/poly_util.cc]: 在编译时，这里的默认参数会报错，可以改为
```
template <class T>
T GetContourArea(std::vector<Point_<T>>& vec, bool oriented ) {
```
也可以尝试其它更好的解决方案，比如参考这里的讨论：https://stackoverflow.com/questions/32105975/why-cant-i-have-template-and-default-arguments

File: paddle/fluid/operators/detection/gpc.cc (hunk 2 above)
Comment (Yipeng-Sun) [reply paddle/fluid/operators/detection/gpc.cc:795]: 涉及FILE *fp的两个函数 xxx_read, xxx_write并没有用到，可以直接注释掉

File: paddle/fluid/operators/detection/poly_util.cc (hunk 3 above)
Comment (Yipeng-Sun) [reply paddle/fluid/operators/detection/poly_util.cc]: 本地环境编译并未出现该问题，默认值已取消

Comment (Yipeng-Sun) [reply paddle/fluid/operators/detection/poly_util.cc]: 更新过上述两个地方后，编译依然正常，仍然出现链接问题；结果无法复现

Comment (Yipeng-Sun) [reply paddle/fluid/operators/detection/poly_util.cc]: [100%] Built target paddle_fluid_shared
[100%] Linking CXX executable test_inference_image_classification_vgg
[100%] Linking CXX executable test_inference_recognize_digits_conv
[100%] Linking CXX executable test_inference_image_classification_resnet
[100%] Linking CXX executable test_inference_word2vec
[100%] Linking CXX executable test_inference_label_semantic_roles
[100%] Linking CXX executable test_inference_recommender_system
[100%] Linking CXX executable test_inference_recognize_digits_mlp
../../libpaddle_fluid.a(multiclass_nms_op.cc.o): In function `paddle::operators::MultiClassNMSKernel<double>::NMSFast(paddle::framework::Tensor const&, paddle::framework::Tensor const&, double, double, double, long, std::vector<int, std::allocator<int> >*) const':
multiclass_nms_op.cc:(.text._ZNK6paddle9operators19MultiClassNMSKernelIdE7NMSFastERKNS_9framework6TensorES6_dddlPSt6vectorIiSaIiEE[_ZNK6paddle9operators19MultiClassNMSKernelIdE7NMSFastERKNS_9framework6TensorES6_dddlPSt6vectorIiSaIiEE]+0x381): undefined reference to `double paddle::operators::PolyArea<double>(double const*, unsigned long, bool)'
multiclass_nms_op.cc:(.text._ZNK6paddle9operators19MultiClassNMSKernelIdE7NMSFastERKNS_9framework6TensorES6_dddlPSt6vectorIiSaIiEE[_ZNK6paddle9operators19MultiClassNMSKernelIdE7NMSFastERKNS_9framework6TensorES6_dddlPSt6vectorIiSaIiEE]+0x399): undefined reference to `double paddle::operators::PolyArea<double>(double const*, unsigned long, bool)'
multiclass_nms_op.cc:(.text._ZNK6paddle9operators19MultiClassNMSKernelIdE7NMSFastERKNS_9framework6TensorES6_dddlPSt6vectorIiSaIiEE[_ZNK6paddle9operators19MultiClassNMSKernelIdE7NMSFastERKNS_9framework6TensorES6_dddlPSt6vectorIiSaIiEE]+0x3b6): undefined reference to `double paddle::operators::PolyOverlapArea<double>(double const*, double const*, unsigned long, bool)'

**Pull request closed without merging**